rfcs/2020-04-20-Optimizer-minimize.md

```diff
@@ -0,0 +1,293 @@
+
+(1) Custom gradient clipping
+
+```python
+optimizer = tf.keras.optimizers.Adam(0.1, transform_gradients=my_gradient_clipping)
+```
+
+(2) Mixed precision (most complicated example):
+
+```python
+class LossScaleOptimizer(Optimizer)
+  def __init__(self, optimizer):
+    self.optimizer = optimizer
+
+  def _get_hyper(self, name):
+    # Optional. Allows access to the wrapped Optimizer's 
+    # hyperparameters (e.g. learning_rate)
+    self.optimizer._get_hyper(name)
+
+  def _transform_loss(self, loss):
+    loss = self.optimizer._transform_loss(loss)
+    # Mixed precision needs to scale loss before calculating gradients
+    return self.scale_loss(loss)
+
+  def _transform_unaggregated_gradients(self, grads_and_vars):
+    # Note: For performance, we could add a check here to see if
+    # self.optimizer._transform_unaggregated_gradients is not implemented, and if
+    # so to skip these scaling / unscalings. Or Grappler could optimize it out.
+    gradients, variables = unpack(grads_and_vars)
+    gradients = self.unscale_gradients(gradients)
+    gradients = self.optimizer._transform_unaggregated_gradients(gradients)
+    # Mixed precision needs to all-reduce on scaled gradients.
+    gradients = self.scale_gradients(gradients)
+    return zip(gradients, variables)
+
+  def _aggregate_gradients(self, grads_and_vars):
+    return aggregate_in_fp16(grads_and_vars)
+
+  def _transform_gradients(self, grads_and_vars):
+    gradients, variables = unpack(grads_and_vars)
+    gradients = unscale_gradients(gradients)
+    gradients = self.optimizer._transform_fgradients(gradients)
+    return zip(gradients, updates)
+
+  def _apply_updates(self, grads_and_vars):
+    return self.optimizer._apply_updates(grads_and_vars)
+```
+
+(3) Horovod (only needs custom aggregation):
+
+To support backwards compatibility for Horovod:
+
+```python
+class HorovodOptimizer(Optimizer):
+  def __init__(self, optimizer):
+    self.optimizer = optimizer
+
+  def _get_hyper(self, name):
+    # See previous example
+    self.optimizer._get_hyper(name)
+
+ def _aggregate_gradients(self, grads_and_vars):
+    return horovod_aggregate_gradients(grads_and_vars)
+
+ # All other methods described in this proposal simply delegate to `self.optimizer`
+```
+    
+Or, if backwards compatibility is not needed, simply:
+
+```python
+optimizer = tf.keras.optimizers.Adam(1e-3, aggregate_gradients=horovod.aggregate)
+```
+
+
+## Alternatives considered
+
+#### Handle this only in Model.fit, via custom hooks exposed on the Model class
+    
+Why rejected:
+
+Shifts the responsibility for implementing and calling these hooks onto each user rather than the writer of the Optimizer subclass (Many users will write custom training logic, many fewer will write Optimizer subclasses).
+
+Solution is too Keras-specific, doesn’t solve the general problem.
+
+
+## Questions and Discussion Topics
+
+Should we create a utility class to help with wrapping an `Optimizer`? I.e. `OptimizerWrapper`?
```

Comment:
As others have said, we should have an `OptimizerWrapper` helper class. In [loss_scale_optimizer.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/mixed_precision/experimental/loss_scale_optimizer.py), about half the code (~300 lines) is to handling the wrapping logic, and all optimizer wrappers will have to have similar code to fully emulate the Optimizer interface. I still plan on adding more wrapping logic in the future to delegate attributes to the inner optimizer. Better to move all that code to a helper class.

Reply:
yes yes and yes :D

Reply:
Sounds good, removed this from the discussion section and gave it a section on the main proposal